resident/resident-service/src/main/java/io/mosip/resident/service/impl/ProxyMasterdataServiceImpl.java

```diff
@@ -413,5 +458,88 @@ public ResponseWrapper<?> getGenderTypesByLangCode(String langCode) throws Resid
 		logger.debug("ProxyMasterdataServiceImpl::getGenderTypesByLangCode()::exit");
 		return responseWrapper;
 	}
+  
+	/**
+	 * download registration centers based on language code, hierarchyLevel and center names
+	 */
+	public byte[] downloadRegistrationCentersByHierarchyLevel(String langCode, Short hierarchyLevel,
+			List<String> name) throws ResidentServiceCheckedException,IOException,Exception {		
+		logger.debug("ResidentServiceImpl::getResidentServicePDF()::entry");
+		ResponseWrapper<?> proxyResponseWrapper = proxyMasterdataService
+				.getAllTemplateBylangCodeAndTemplateTypeCode(langCode, REGISTRATION_CENTER_TEMPLATE_NAME);		
+		ResponseWrapper<?> regCentResponseWrapper = proxyMasterdataService.getRegistrationCentersByHierarchyLevel(langCode,hierarchyLevel, name);
+		Map<String,Object> regCentersMap = new LinkedHashMap<>();
+		List<RegistrationCenterDto >regCentersDtlsList=Collections.EMPTY_LIST;
+		int serialNumber = 1;
+		if (regCentResponseWrapper != null) {
+			RegistrationCenterResponseDto registrationCentersDtls = mapper.readValue(
+					mapper.writeValueAsString(regCentResponseWrapper.getResponse()), RegistrationCenterResponseDto.class);
+			regCentersDtlsList = registrationCentersDtls.getRegistrationCenters();
+			for (RegistrationCenterDto regCenterDto : regCentersDtlsList) {
+				List<WorkingDaysDto> workingDaysList = Collections.EMPTY_LIST;
+				String workingHours = "";
+				String fullAddress = getFullAddress(regCenterDto.getAddressLine1(), regCenterDto.getAddressLine2(),
+						regCenterDto.getAddressLine3());
+				regCenterDto.setSerialNumber(serialNumber++);
+				regCenterDto.setFullAddress(fullAddress);				
+				workingDaysList = getRegCenterWorkingDays(regCenterDto.getId(), regCenterDto.getLangCode());
+				workingHours = workingDaysList.get(0).getName() + "-" + workingDaysList.get(1).getName() + "|"
+						+ regCenterDto.getCenterStartTime().substring(0, 5) + " " + "am" + "-"
+						+ regCenterDto.getCenterEndTime().substring(0, 5) + " " + "pm";
+				regCenterDto.setWorkingHours(workingHours);
+			}
+		}				
+		regCentersMap.put("regCentersDtlsList", regCentersDtlsList);
+		logger.debug("template data from DB:" + proxyResponseWrapper.getResponse());
+		Map<String, Object> templateResponse = new LinkedHashMap<>((Map<String, Object>) proxyResponseWrapper.getResponse());
+		String fileText = (String) templateResponse.get("fileText");
+		InputStream downLoadRegCenterTemplate = new ByteArrayInputStream(fileText.getBytes(StandardCharsets.UTF_8));
+		InputStream downLoadRegCenterTemplateData = templateManager.merge(downLoadRegCenterTemplate, regCentersMap);
+		
+		StringWriter writer = new StringWriter();
+		IOUtils.copy(downLoadRegCenterTemplateData, writer, "UTF-8");
+		ByteArrayOutputStream pdfValue = (ByteArrayOutputStream) pdfGenerator.generate(writer.toString());
+		logger.debug("ResidentServiceImpl::residentServiceHistoryPDF()::exit");
+		return pdfValue.toByteArray();
```

Comment:
This is dangerous. We can not be doing this. Please look at webflux or spring async. How many calls are expected? this many byte arrays will kill our system.

Reply:
modified the code as per the review comments and removed byte arrays